Tools/clinic/clinic.py

```diff
@@ -2044,10 +2044,16 @@ def write_file(filename: str, new_contents: str) -> None:
         raise
 
 
+class Parser:
```

Comment:
Consider having this class inherit from `typing.Protocol`. If you do so, then mypy will understand that `PythonParser` and `DSLParser` both implement the `Parser` interface, even if they don't inherit from the `Parser` class: https://docs.python.org/3/library/typing.html#typing.Protocol.

Protocols allow you to use "static duck typing", which imo is a more Pythonic solution to this problem than an ABC

Reply:
All right, I'll get to it later today. Thanks!

Reply:
Since this is an app, not a library, you could also consider doing away with this `Parser` class entirely, and changing the definition of `ParserDict` to be `ParserDict = dict[str, "PythonParser | DSLParser"]`. It would be _slightly_ less principled, but it would also be simpler and less verbose. It might honestly work just as well for our use case here.

But also, I'm fine with keeping the `Parser` class (if you make it a Protocol ;)

Reply:
Hm, if I remove the Parser class and annotate `ParserDict` more verbosely, mypy complains:

```
Tools/clinic/clinic.py:2216: error: Incompatible types in assignment (expression has type "object", target has type "PythonParser | DSLParser")  [assignment]
                        self.parsers[dsl_name] = parsers[dsl_name](self)
                                                 ^~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 2 source files)
```

Perhaps the `Protocol` way is better.

Reply:
> Perhaps the `Protocol` way is better.

Go for it!

Reply:
Aaaah, the toplevel `parser` object and `Clinic.parser` are slightly different dicts. The former holds the parser types, the latter the parser instances. Annotating them both with `ParserDict` is incorrect.

Reply:
The treachery of type aliases strikes again ;)

Reply:
I'm pretty satisfied with the current solution; thanks! :)